src/jsdoc_transformer.ts

```diff
@@ -112,88 +112,107 @@ export function maybeAddHeritageClauses(
+        // It's not clear when this can happen.
+        mtt.debugWarn(decl, `could not get type of symbol: ${expr.getText()}`);
+        return null;
       }
-      const typeTranslator = mtt.newTypeTranslator(impl.expression);
-      if (typeTranslator.isBlackListed(alias)) {
-        continue;
+      alias = type.symbol;
+    }
+    if (alias.flags & ts.SymbolFlags.Alias) {
+      alias = mtt.typeChecker.getAliasedSymbol(alias);
+    }
+    const typeTranslator = mtt.newTypeTranslator(expr.expression);
+    if (typeTranslator.isBlackListed(alias)) {
+      // Don't emit references to blacklisted types.
+      return null;
+    }
+
+    if (alias.flags & ts.SymbolFlags.Class) {
+      if (!isClass) {
+        // Closure interfaces cannot extend or implements classes.
+        mtt.debugWarn(decl, `omitting interface deriving from class: ${expr.getText()}`);
+        return null;
       }
-      // We can only @implements an interface, not a class.
-      // But it's fine to translate TS "implements Class" into Closure
-      // "@extends {Class}" because this is just a type hint.
-      if (alias.flags & ts.SymbolFlags.Class) {
-        if (!isClass) {
-          // Only classes can extend classes in TS. Ignoring the heritage clause should be safe,
-          // as interfaces are @record anyway, so should prevent property disambiguation.
-
-          // Problem: validate that methods are there?
-          continue;
-        }
-        if (classHasSuperClass && heritage.token !== ts.SyntaxKind.ExtendsKeyword) {
-          // Do not emit an @extends for a class that already has a proper ES6 extends class. This
-          // risks incorrect optimization, as @extends takes precedence, and Closure won't be
-          // aware of the actual type hierarchy of the class.
-          continue;
-        }
-        tagName = 'extends';
-      } else if (alias.flags & ts.SymbolFlags.Value) {
-        // If the symbol was already in the value namespace, then it will
-        // not be a type in the Closure output (because Closure collapses
-        // the type and value namespaces).  Just ignore the implements.
-        continue;
+      if (!isExtends) {
+        // Closure can only @implements an interface, not a class.
+        mtt.debugWarn(decl, `omitting @implements of a class: ${expr.getText()}`);
+        return null;
       }
-      // typeToClosure includes nullability modifiers, so call symbolToString directly here.
-      docTags.push({tagName, type: typeTranslator.symbolToString(alias, true)});
+    } else if (alias.flags & ts.SymbolFlags.Value) {
+      // If it's something other than a classin the value namespace, then it will
+      // not be a type in the Closure output (because Closure collapses
+      // the type and value namespaces).
+      mtt.debugWarn(
+          decl, `omitting heritage reference to a type/value conflict: ${expr.getText()}`);
+      return null;
+    } else if (alias.flags & ts.SymbolFlags.TypeLiteral) {
```

Comment:
I wonder if more generally, we should be whitelisting here instead of blacklisting known bad situations. Not sure if it's worth it.